mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala

```diff
@@ -0,0 +1,235 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.regression
+
+import org.apache.spark.api.java.{JavaRDD, JavaPairRDD}
+import org.apache.spark.rdd.RDD
+
+/**
+ * Regression model for Isotonic regression
+ *
+ * @param predictions Weights computed for every feature.
+ * @param isotonic isotonic (increasing) or antitonic (decreasing) sequence
+ */
+class IsotonicRegressionModel (
+    val predictions: Seq[(Double, Double, Double)],
+    val isotonic: Boolean)
+  extends Serializable {
+
+  /**
+   * Predict labels for provided features
+   *
+   * @param testData features to be labeled
+   * @return predicted labels
+   */
+  def predict(testData: RDD[Double]): RDD[Double] =
+    testData.map(predict)
+
+  /**
+   * Predict labels for provided features
+   *
+   * @param testData features to be labeled
+   * @return predicted labels
+   */
+  def predict(testData: JavaRDD[java.lang.Double]): JavaRDD[java.lang.Double] =
+    testData.rdd.map(_.doubleValue()).map(predict).map(new java.lang.Double(_))
+
+  /**
+   * Predict a single label
+   *
+   * @param testData feature to be labeled
+   * @return predicted label
+   */
+  def predict(testData: Double): Double =
+    // Take the highest of data points smaller than our feature or data point with lowest feature
+    (predictions.head +: predictions.filter(y => y._2 <= testData)).last._1
+}
+
+/**
+ * Base representing algorithm for isotonic regression
+ */
+trait IsotonicRegressionAlgorithm
```

Comment:
We can do this later when we have multiple implementations for isotonic regression. We try to ask users to use builder pattern instead of static methods to train models, because the latter approach is hard to maintain. For this PR, it might be easier to expose `PoolAdjacentViolators` as `IsotonicRegression` and remove this trait.